endpoints/s3/src/completemultipartupload.cpp

```diff
@@ -0,0 +1,418 @@
+#include "irods/private/s3_api/s3_api.hpp"
+#include "irods/private/s3_api/authentication.hpp"
+#include "irods/private/s3_api/bucket.hpp"
+#include "irods/private/s3_api/common_routines.hpp"
+#include "irods/private/s3_api/connection.hpp"
+#include "irods/private/s3_api/log.hpp"
+#include "irods/private/s3_api/common.hpp"
+#include "irods/private/s3_api/session.hpp"
+#include "irods/private/s3_api/configuration.hpp"
+#include "irods/private/s3_api/globals.hpp"
+
+#include <irods/dstream.hpp>
+#include <irods/transport/default_transport.hpp>
+#include <irods/irods_exception.hpp>
+#include <irods/irods_at_scope_exit.hpp>
+
+#include <boost/uuid/uuid.hpp>
+#include <boost/uuid/uuid_generators.hpp>
+#include <boost/uuid/uuid_io.hpp>
+#include <boost/lexical_cast.hpp>
+#include <boost/property_tree/ptree.hpp>
+#include <boost/property_tree/xml_parser.hpp>
+
+#include <fmt/format.h>
+#include <regex>
+#include <cstdio>
+#include <vector>
+#include <mutex>
+#include <condition_variable>
+#include <memory>
+#include <thread>
+#include <chrono>
+#include <string>
+#include <sstream>
+
+namespace asio = boost::asio;
+namespace beast = boost::beast;
+namespace fs = irods::experimental::filesystem;
+namespace logging = irods::http::logging;
+
+namespace {
+
+    std::regex upload_id_pattern("[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}");
+
+    // store offsets and lengths for each part
+    struct part_info_t {
```

Comment:
Remove `_t` from type name.

That is normally a convention that's used with `typedef`.

Reply:
done